marathon/src/main/scala/io/buoyant/marathon/v2/AppIdNamer.scala

```diff
@@ -37,93 +35,99 @@ class AppIdNamer(
             Trace.recordBinary("marathon.appId", app.show)
-            Trace.recordBinary("marathon.id", id.show)
-            Trace.recordBinary("marathon.found", app.show)
-
-            val addr = getAndMonitorAddr(app.show)
-            NameTree.Leaf(Name.Bound(addr, id, residual))
-
-          case None =>
-            Trace.recordBinary("marathon.notfound", path.show)
-            NameTree.Neg
+            val residual = path.drop(app.size)
+            val id = prefix ++ app
+            val addr = getAndMonitorAddr(app)
+            Name.Bound(addr, id, residual)
+        } match {
+          case None => NameTree.Neg
+          case Some(name) => NameTree.Leaf(name)
         }
+      }
     }
 
-  private[this] val appsActivity: Activity[Api.AppIds] = {
-    val states = Var.async[Activity.State[Api.AppIds]](Activity.Pending) { state =>
-      def loop(): Future[Unit] =
-        api.getAppIds().transform {
+  private[this] val appsActivity: Activity[Api.AppIds] =
+    Activity(Var.async[Activity.State[Api.AppIds]](Activity.Pending) { state =>
+      @volatile var initialized, stopped = false
+      @volatile var pending: Future[_] = Future.never
+
+      def loop(): Unit = if (!stopped) {
+        pending = api.getAppIds().respond {
           case Return(apps) =>
+            initialized = true
             state() = Activity.Ok(apps)
-            Future.sleep(ttl).before(loop())
+            if (!stopped) Future.sleep(ttl).map(_ => loop())
 
           case Throw(NonFatal(e)) =>
-            state() = Activity.Failed(e)
-            Future.sleep(ttl).before(loop())
+            if (!initialized) {
+              state() = Activity.Failed(e)
+            }
+            if (!stopped) Future.sleep(ttl).map(_ => loop())
 
           case Throw(e) =>
             state() = Activity.Failed(e)
-            Future.exception(e)
         }
+      }
 
-      val work = loop()
+      loop()
       Closable.make { deadline =>
-        work.raise(Closed)
+        stopped = true
+        pending.raise(Closed)
         Future.Unit
       }
-    }
-    Activity(states)
-  }
+    })
 
-  private[this] var appMonitors: Map[String, Var[Addr]] = Map.empty
-  private[this] def getAndMonitorAddr(app: String): Var[Addr] = synchronized {
+  private[this] var appMonitors: Map[Path, Var[Addr]] = Map.empty
+
+  private[this] def getAndMonitorAddr(app: Path): Var[Addr] = synchronized {
     appMonitors.get(app) match {
       case Some(addr) => addr
 
       case None =>
         val addr = Var.async[Addr](Addr.Pending) { addr =>
-          def loop(): Future[Unit] =
-            api.getAddrs(app).transform { ret =>
-              ret match {
-                case Return(addrs) =>
-                  addr() = Addr.Bound(addrs)
-                  Future.sleep(ttl).before(loop())
-
-                case Throw(NonFatal(e)) =>
+          @volatile var initialized, stopped = false
+          @volatile var pending: Future[_] = Future.never
```

Comment:
yeah, it's there only so that if the Var is closed we can tear down a pending request.